cpp/src/arrow/scalar.h

```diff
@@ -237,19 +246,17 @@ struct ARROW_EXPORT FixedSizeBinaryScalar : public BinaryScalar {
   explicit FixedSizeBinaryScalar(std::shared_ptr<DataType> type) : BinaryScalar(type) {}
 };
 
-template <typename T>
-struct ARROW_EXPORT TemporalScalar : public Scalar {
-  using Scalar::Scalar;
+template <typename T, typename StorageType = typename T::StorageType,
+          typename Enable = void>
+struct ARROW_EXPORT TemporalScalar : internal::PrimitiveScalar<StorageType> {
```

Comment:
I'm not sure this use of inheritance is kosher. This implies `TimestampScalar isa Int64Scalar`, which is a relationship not present in parallel hierarchies (Array, DataType, Builder) and is therefore likely to be confusing. Why not use `TemporalScalar : PrimitiveScalar<T>`?

Reply:
Int64Scalar and TimestampScalar need to have a common base type in order to both be unboxed by templated code that uses Int64Type in the kernel codegen. So `PrimitiveScalar<TimestampType>` won't work.